File: bindings/typescript/src/algorithms/paths.ts
// Path-finding algorithms as pure functions.
//
// This is the **most inconsistent** algorithm module and the asymmetries are
// reproduced verbatim from bindings/python/src/graphqlite/algorithms/paths.py:
//   1. `shortestPath` quotes with " ; `astar` quotes with ' .
//   2. Both `shortestPath` and `astar` unwrap the `column_0` wrapper that this
//      core wraps results in, then fall back to direct `result[0]` field access
//      when it is absent. (Historically `astar` skipped the unwrap and always
//      returned defaults; fixed in #64.)
//   3. `astar`'s lat/lon props are interpolated unescaped in Python; here they
//      are validated with `assertIdentifier` instead.
//   4. Only `apsp` goes through `.toList()` → `extractAlgoArray` (which unwraps
//      the `column_0` array, so apsp works).
import type { Connection } from '../connection.ts';
import { escapeString, assertIdentifier } from '../utils.ts';
import { extractAlgoArray, safeFloat, safeInt } from './parsing.ts';

/** `{ path, distance, found }` — a single shortest path. */
export interface ShortestPathResult {
  path: unknown[];
  distance: number | null;
  found: boolean;
}

/** `shortestPath` result plus A*'s `nodesExplored` counter. */
export interface AStarResult extends ShortestPathResult {
  nodesExplored: number;
}

/** `{ source, target, distance }` — one reachable pair from all-pairs shortest path. */
export interface AllPairsPath {
  source: unknown;
  target: unknown;
  distance: number;
}

export interface ShortestPathOptions {
  /** Edge property used as weight; omitted → unweighted (hop count). */
  weightProp?: string;
}

export interface AStarOptions {
  /** Latitude/y coordinate property (interpolated → validated as an identifier). */
  latProp?: string;
  /** Longitude/x coordinate property (interpolated → validated as an identifier). */
  lonProp?: string;
}

/**
 * Dijkstra shortest path. **Uses double quotes** and **unwraps `column_0`** —
 * both unique to this method (see module header). Empty result → the documented
 * `{ path: [], distance: null, found: false }` default.
 */
export function shortestPath(
  conn: Connection,
  sourceId: string,
  targetId: string,
  options: ShortestPathOptions = {},
): ShortestPathResult {
  const escSource = escapeString(sourceId);
  const escTarget = escapeString(targetId);
  const query = options.weightProp
    ? `RETURN dijkstra("${escSource}", "${escTarget}", "${escapeString(options.weightProp)}")`
    : `RETURN dijkstra("${escSource}", "${escTarget}")`;

  const result = conn.cypher(query);
  if (result.length === 0) {
    return { path: [], distance: null, found: false };
  }

  const row = result[0]!;
  const col0 = row['column_0'];
  if (col0 && typeof col0 === 'object' && !Array.isArray(col0)) {
    const data = col0 as Record<string, unknown>;
    return {
      path: (data['path'] as unknown[]) ?? [],
      distance: (data['distance'] as number | null) ?? null,
      found: (data['found'] as boolean) ?? false,
    };
  }
  // Already unpacked (defensive — Python's direct-access branch).
  return {
    path: (row['path'] as unknown[]) ?? [],
    distance: (row['distance'] as number | null) ?? null,
    found: (row['found'] as boolean) ?? false,
  };
}

/** Alias for {@link shortestPath} (matches Python's `dijkstra`). */
export const dijkstra = shortestPath;

/**
 * A* shortest path. **Uses single quotes** and **unwraps `column_0`** (like
 * {@link shortestPath}), reading `path`/`distance`/`found`/`nodes_explored` off
 * the unwrapped object, with a defensive fallback to direct `result[0]` access.
 * `latProp`/`lonProp` are interpolated, so they are validated with
 * {@link assertIdentifier}. Empty result → the documented default (with
 * `nodesExplored: 0`).
 */
export function astar(
  conn: Connection,
  sourceId: string,
  targetId: string,
  options: AStarOptions = {},
): AStarResult {
  const escSource = escapeString(sourceId);
  const escTarget = escapeString(targetId);

  let query: string;
  if (options.latProp && options.lonProp) {
    assertIdentifier(options.latProp, 'property');
    assertIdentifier(options.lonProp, 'property');
    query = `RETURN astar('${escSource}', '${escTarget}', '${options.latProp}', '${options.lonProp}')`;
  } else {
    query = `RETURN astar('${escSource}', '${escTarget}')`;
  }

  const result = conn.cypher(query);
  if (result.length === 0) {
    return { path: [], distance: null, found: false, nodesExplored: 0 };
  }

  const row = result[0]!;
  const col0 = row['column_0'];
  if (col0 && typeof col0 === 'object' && !Array.isArray(col0)) {
    const data = col0 as Record<string, unknown>;
    return {
      path: (data['path'] as unknown[]) ?? [],
      distance: (data['distance'] as number | null) ?? null,
      found: (data['found'] as boolean) ?? false,
      nodesExplored: safeInt(data['nodes_explored']),
    };
  }
  // Already unpacked (defensive — Python's direct-access branch).
  return {
    path: (row['path'] as unknown[]) ?? [],
    distance: (row['distance'] as number | null) ?? null,
    found: (row['found'] as boolean) ?? false,
    nodesExplored: safeInt(row['nodes_explored']),
  };
}

/** Alias for {@link astar} (matches Python's `a_star`). */
export const aStar = astar;

/**
 * All-pairs shortest path (Floyd–Warshall). The **only** path method that runs
 * through `.toList()` → {@link extractAlgoArray} (which unwraps `column_0`).
 * Keeps only rows with non-null `source` and `target`.
 */
export function allPairsShortestPath(conn: Connection): AllPairsPath[] {
  const rows = extractAlgoArray(conn.cypher('RETURN apsp()').toList());
  const paths: AllPairsPath[] = [];
  for (const row of rows) {
    const source = row['source'];
    const target = row['target'];
    if (source != null && target != null) {
      paths.push({ source, target, distance: safeFloat(row['distance']) });
    }
  }
  return paths;
}

/** Alias for {@link allPairsShortestPath} (matches Python's `apsp`). */
export const apsp = allPairsShortestPath;

File: bindings/rust/src/algorithms/paths.rs
//! Path finding algorithm implementations.

use super::parsing::{algo_object, algo_rows, extract_float, extract_string};
use super::{AStarResult, ApspResult, ShortestPathResult};
use crate::graph::Graph;
use crate::utils::{assert_identifier, escape_string};
use crate::{Result, Value};
use std::collections::HashMap;

fn path_field(data: &HashMap<String, Value>) -> Vec<String> {
    match data.get("path") {
        Some(Value::Array(arr)) => arr
            .iter()
            .filter_map(|v| match v {
                Value::String(s) => Some(s.clone()),
                _ => None,
            })
            .collect(),
        _ => Vec::new(),
    }
}

fn distance_field(data: &HashMap<String, Value>) -> Option<f64> {
    data.get("distance").and_then(Value::as_f64)
}

fn found_field(data: &HashMap<String, Value>) -> bool {
    data.get("found").and_then(Value::as_bool).unwrap_or(false)
}

impl Graph {
    /// Find the shortest path between two nodes using Dijkstra's algorithm.
    ///
    /// # Arguments
    ///
    /// * `source_id` - ID of the source node
    /// * `target_id` - ID of the target node
    /// * `weight_property` - Optional edge property to use as weight
    pub fn shortest_path(
        &self,
        source_id: &str,
        target_id: &str,
        weight_property: Option<&str>,
    ) -> Result<ShortestPathResult> {
        let esc_source = escape_string(source_id);
        let esc_target = escape_string(target_id);

        let query = match weight_property {
            Some(wp) => format!(
                "RETURN dijkstra(\"{}\", \"{}\", \"{}\")",
                esc_source,
                esc_target,
                escape_string(wp)
            ),
            None => format!("RETURN dijkstra(\"{}\", \"{}\")", esc_source, esc_target),
        };

        let result = self.connection().cypher(&query)?;

        Ok(match algo_object(&result) {
            Some(data) => ShortestPathResult {
                path: path_field(&data),
                distance: distance_field(&data),
                found: found_field(&data),
            },
            None => ShortestPathResult {
                path: Vec::new(),
                distance: None,
                found: false,
            },
        })
    }

    /// Find shortest path using A* algorithm with heuristic guidance.
    ///
    /// # Arguments
    ///
    /// * `source_id` - Starting node's id
    /// * `target_id` - Target node's id
    /// * `lat_prop` - Optional property name for latitude (must be a plain identifier)
    /// * `lon_prop` - Optional property name for longitude (must be a plain identifier)
    ///
    /// # Errors
    ///
    /// Returns [`crate::Error::InvalidIdentifier`] if a coordinate property
    /// name is not a valid identifier; nothing is executed in that case.
    pub fn astar(
        &self,
        source_id: &str,
        target_id: &str,
        lat_prop: Option<&str>,
        lon_prop: Option<&str>,
    ) -> Result<AStarResult> {
        let esc_source = escape_string(source_id);
        let esc_target = escape_string(target_id);

        let query = match (lat_prop, lon_prop) {
            (Some(lat), Some(lon)) => {
                assert_identifier(lat)?;
                assert_identifier(lon)?;
                format!(
                    "RETURN astar('{}', '{}', '{}', '{}')",
                    esc_source, esc_target, lat, lon
                )
            }
            _ => format!("RETURN astar('{}', '{}')", esc_source, esc_target),
        };

        let result = self.connection().cypher(&query)?;

        Ok(match algo_object(&result) {
            Some(data) => AStarResult {
                path: path_field(&data),
                distance: distance_field(&data),
                found: found_field(&data),
                nodes_explored: data
                    .get("nodes_explored")
                    .and_then(Value::as_i64)
                    .unwrap_or(0),
            },
            None => AStarResult {
                path: Vec::new(),
                distance: None,
                found: false,
                nodes_explored: 0,
            },
        })
    }

    /// Compute shortest paths between all pairs of nodes.
    ///
    /// Uses Floyd-Warshall algorithm with O(V³) time complexity.
    pub fn apsp(&self) -> Result<Vec<ApspResult>> {
        let result = self.connection().cypher("RETURN apsp()")?;
        let rows = algo_rows(&result);

        let mut paths = Vec::new();
        for row in rows.iter() {
            if let (Some(source), Some(target)) =
                (extract_string(row, "source"), extract_string(row, "target"))
            {
                paths.push(ApspResult {
                    source,
                    target,
                    distance: extract_float(row, "distance"),
                });
            }
        }
        Ok(paths)
    }
}
